homeassistant/components/mqtt/mixins.py

```diff
@@ -1120,7 +1121,22 @@ def _setup_common_attributes_from_config(self, config: ConfigType) -> None:
             config.get(CONF_ENABLED_BY_DEFAULT)
         )
         self._attr_icon = config.get(CONF_ICON)
-        self._attr_name = config.get(CONF_NAME)
+        entity_name: str | None = config.get(CONF_NAME)
+        if CONF_DEVICE in config:
+            device_name: str | None = config[CONF_DEVICE].get(CONF_NAME)
+            if device_name is not None and entity_name is None:
+                self._attr_name = None
+                self._attr_has_entity_name = True
+                return
+            if device_name is not None and entity_name is not None:
+                self._attr_name = entity_name
+                self._attr_has_entity_name = True
+                return
```

Comment:
What's the point of two if statements, it seems like in both cases `self._attr_name` will be set to `entity_name`? 

Also, the early returns are not great if we want to extend this in a future. Move this code to a new helper instead.